File: week05/5/prob1.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* removeElements(ListNode* head, int val) {
        ListNode* sentinel = new ListNode(-1, nullptr);
        ListNode* prev = sentinel;
        ListNode* curr = head;

        while (curr) {
            if (curr->val == val) {
                prev->next = nullptr;
                // prev = prev
                curr = curr->next;
            } else {
                prev->next = curr;
                prev = curr;
                curr = curr->next;
            }
        }
        return sentinel->next;
    }
};

File: week05/5/prob1.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeElements(self, head: Optional[ListNode], val: int) -> Optional[ListNode]:
        sentinel = ListNode(-1, None)
        prev = sentinel
        curr = head
        while curr:
            if curr.val == val:
                prev.next = None
                # prev = prev
                curr = curr.next
            else:
                prev.next = curr
                prev = curr
                curr = curr.next
        return sentinel.next
